File: scan/hoard-scan/Sources/ScanLink/Pairing.swift
import CryptoKit
import Foundation
import Network

public struct PairingCode: Equatable, Sendable {
    public let digits: String

    public init?(_ raw: String) {
        let cleaned = raw.filter(\.isNumber)
        guard cleaned.count == 6 else { return nil }
        digits = cleaned
    }

    public static func random() -> PairingCode {
        var g = SystemRandomNumberGenerator()
        let n = UInt32.random(in: 0..<1_000_000, using: &g)
        return PairingCode(String(format: "%06u", n))!
    }

    public var display: String {
        let mid = digits.index(digits.startIndex, offsetBy: 3)
        return "\(digits[digits.startIndex..<mid]) \(digits[mid...])"
    }

    var key: SymmetricKey {
        let salt = Data("dev.spiffcs.hoard.scan.pairing.v1".utf8)
        return HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: Data(digits.utf8)),
            salt: salt,
            outputByteCount: 32)
    }
}

func proof(session: String, code: PairingCode, peerFingerprint: Data? = nil) -> String {
    var mac = HMAC<SHA256>(key: code.key)
    mac.update(data: Data(session.utf8))
    if let peerFingerprint {
        mac.update(data: Data([0x00]))
        mac.update(data: peerFingerprint)
    }
    return Data(mac.finalize()).base64EncodedString()
}

func verifyProof(
    _ claimed: String, session: String, code: PairingCode, ownFingerprint: Data? = nil
) -> Bool {
    guard let bytes = Data(base64Encoded: claimed) else { return false }
    var authenticated = Data(session.utf8)
    if let ownFingerprint {
        authenticated.append(0x00)
        authenticated.append(ownFingerprint)
    }
    return HMAC<SHA256>.isValidAuthenticationCode(
        bytes, authenticating: authenticated, using: code.key)
}

public let scanServiceType = "_hoardscan._tcp"

public enum PeerRole: String, Codable, Sendable {
    case control
    case preview
}

struct PeerHello: Codable {
    var role: PeerRole
    var session: String
    var proof: String
    var name: String = ""
}

File: internal/scan/link/pairing.go
package link

import (
	"crypto/hkdf"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"fmt"
)

const ServiceType = "_hoardscan._tcp"

const pairingSalt = "dev.spiffcs.hoard.scan.pairing.v1"

const derivedKeyLen = 32

var ErrBadCode = errors.New("link: pairing code must be six digits")

type Code struct {
	digits string
}

func ParseCode(raw string) (Code, error) {
	cleaned := make([]byte, 0, len(raw))
	for i := 0; i < len(raw); i++ {
		if raw[i] >= '0' && raw[i] <= '9' {
			cleaned = append(cleaned, raw[i])
		}
	}
	if len(cleaned) != 6 {
		return Code{}, fmt.Errorf("%w: got %d", ErrBadCode, len(cleaned))
	}
	return Code{digits: string(cleaned)}, nil
}

func (c Code) Digits() string { return c.digits }

func (c Code) Display() string {
	if len(c.digits) != 6 {
		return c.digits
	}
	return c.digits[:3] + " " + c.digits[3:]
}

func (c Code) IsZero() bool { return c.digits == "" }

func (c Code) key() ([]byte, error) {
	return hkdf.Key(sha256.New, []byte(c.digits), []byte(pairingSalt), "", derivedKeyLen)
}

func Proof(session string, c Code, peerFingerprint []byte) (string, error) {
	key, err := c.key()
	if err != nil {
		return "", err
	}
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(session))
	if peerFingerprint != nil {

		mac.Write([]byte{0x00})
		mac.Write(peerFingerprint)
	}
	return base64.StdEncoding.EncodeToString(mac.Sum(nil)), nil
}

func VerifyProof(claimed, session string, c Code, ownFingerprint []byte) bool {
	want, err := Proof(session, c, ownFingerprint)
	if err != nil {
		return false
	}

	return hmac.Equal([]byte(claimed), []byte(want))
}

type Role string

const (
	RoleControl Role = "control"
	RolePreview Role = "preview"
)

type Hello struct {
	Role    Role   `json:"role"`
	Session string `json:"session"`
	Proof   string `json:"proof"`
	Name    string `json:"name"`
}
